src/base/libmesh.C

```diff
@@ -798,7 +816,20 @@ LibMeshInit::~LibMeshInit()
 #endif
 
       if (libmesh_initialized_mpi)
-        MPI_Finalize();
+        {
+          // We can't just libmesh_assert here because destructor,
+          // but we ought to report any errors
+          unsigned int error_code = MPI_Finalize();
```

Comment:
Interestingly, it looks like [some MPI implementations might even have a Barrier in Finalize](http://www.mcs.anl.gov/research/projects/mpi/mpi-standard/mpi-report-2.0/node32.htm) if they are following this implementation advice.
```
An implementation may need to delay the return from MPI_FINALIZE until all potential future message
cancellations have been processed. One possible solution is to place a barrier inside MPI_FINALIZE
```
I wonder if the different behavior you observed stemmed from something like this...